pos_blstm/blstm.py

```diff
@@ -0,0 +1,206 @@
+"""Implements the long-short term memory character model.
+This version vectorizes over multiple examples, but each string
+has a fixed length."""
+
+from __future__ import absolute_import
+from __future__ import print_function
+from builtins import range
+from os.path import dirname, join
+import autograd.numpy as np
+import autograd.numpy.random as npr
+from autograd import grad
+from autograd.scipy.misc import logsumexp
+import sys
+from optimizers import adam
+from rnn import string_to_one_hot, one_hot_to_string,\
+                build_dataset, sigmoid, concat_and_multiply
+import pdb
+import cPickle as pickle
+
+def init_lstm_params(input_size, state_size, output_size,
+                     param_scale=0.01, rs=npr.RandomState(0)):
+    # state_size is length of h vector, input_size is length of x vector
+    # output_size is length of y vector?
+    def rp(*shape): # returns a np array sampled from normal distribution of specified shape
+        return rs.randn(*shape) * param_scale
+
+    if load:
+        return dict(np.load('tmp/trained_weights.npz'))
+    return {'capitals':     rp(3, input_size),
+            'init cells1':   rp(1, state_size),
+            'init hiddens1': rp(1, state_size),
+            'change1':       rp(input_size + state_size + 1, state_size),
+            'forget1':       rp(input_size + state_size + 1, state_size),
+            'ingate1':       rp(input_size + state_size + 1, state_size),
+            'outgate1':      rp(input_size + state_size + 1, state_size),
+            'init cells2':   rp(1, state_size),
+            'init hiddens2': rp(1, state_size),
+            'change2':       rp(input_size + state_size + 1, state_size),
+            'forget2':       rp(input_size + state_size + 1, state_size),
+            'ingate2':       rp(input_size + state_size + 1, state_size),
+            'outgate2':      rp(input_size + state_size + 1, state_size),
+            'predict':      rp(state_size*2 + 1, output_size)}
```

Comment:
maybe call these ingate_f and ingate_b so we don't get confused which one is forward and which is backward.  Nice job adding the bias vectors (i usually forget those)